File: LoadDataFromAPIObjC/LoadDataFromAPIObjC/Services/NetworkHelper.m
//
//  NetworkHelper.m
//  LoadDataFromAPIObjC
//
//  Created by Lisa J on 5/24/18.
//  Copyright © 2018 Lisa J. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "NetworkHelper.h"

//class extension
@interface NetworkHelper()
//private properties & methods
@property (nonatomic) NSURLSession *urlSession;
@end

@implementation NetworkHelper

+(instancetype)sharedManager {
    static NetworkHelper *networkHelper;
    static dispatch_once_t once_token;
    //ensures there's only one instance
    dispatch_once(&once_token, ^ {
        networkHelper = [[NetworkHelper alloc] init];
    });
    return networkHelper;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        //initialize properties
        _urlSession = [NSURLSession sessionWithConfiguration:NSURLSessionConfiguration.defaultSessionConfiguration];
    }
    return self;
}

- (void)performRequestWithRequest:(NSURLRequest *)request completionHandler:(void (^)(NSError *, NSData *))completion {
    NSURLSessionDataTask *dataTask = [self.urlSession
                                      dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                          if (error) {
                                              completion(error,nil);
                                          } else {
                                              completion(nil,data);
                                          }
                                      }];
    [dataTask resume];
}

@end

File: URLSessionNasa/URLSessionNasa/NetworkHelper.swift
//
//  NetworkHelper.swift
//  URLSessionNasa
//
//  Created by Lisa J on 12/1/17.
//  Copyright © 2017 Lisa J. All rights reserved.
//

import Foundation

class NetworkHelper {
    private init() {}
    static let manager = NetworkHelper()
    private let mySession = URLSession(configuration: .default)
    func performDataTask(with url: URL, completionHandler:@escaping (Data) -> Void, errorHandler: @escaping (Error) -> Void) {
        mySession.dataTask(with: url){(data: Data?, response: URLResponse?, error: Error?) in
            guard let data = data else {return}
            DispatchQueue.main.async {
            if let error = error {
                errorHandler(error)
                return
                }
                completionHandler(data)
            }
        }.resume()//needs this to start the data task
        
    }
}
